relay-sampling/src/lib.rs

```diff
@@ -355,6 +365,125 @@ impl TimeRange {
     }
 }
 
+/// A struct representing the execution context of a decaying function.
+#[derive(Clone, Copy, Debug)]
+struct DecayingFunctionExecutor {
+    decaying_fn: DecayingFunction,
+    external_params: DecayingFunctionExternalParams,
+}
+
+impl DecayingFunctionExecutor {
+    fn execute(&self) -> f64 {
+        match self.decaying_fn {
+            DecayingFunction::Linear { from_sample_rate } => {
+                self.execute_linear_decay(from_sample_rate)
+            }
+            DecayingFunction::Constant => self.external_params.base_sample_rate,
+        }
+    }
+
+    /// Computes the linear decay function at a specific point in time.
+    ///
+    /// The linear decay works in the following way:
+    /// start = 100
+    /// end = 200
+    /// from = 0.7 (70% sample rate)
+    /// base_sample_rate = 0.5 (50% sample rate)
+    /// (from - base_sample_rate) = 0.2 (20% sample rate)
+    ///
+    /// If we have now = 170 it means we are at the 70% progress of the timerange, which means we
+    /// want to inversely reduce the sample rate in the interval (from - base_sample_rate).
+    /// Thus, in the 70% case, we would take only 30% of the (from - base_sample_rate) which
+    /// is equal to 0.06.
+    fn execute_linear_decay(&self, from_sample_rate: f64) -> f64 {
+        let now_timestamp = self.external_params.now.timestamp() as f64;
+        let start_timestamp = self.external_params.start.timestamp() as f64;
+        let end_timestamp = self.external_params.end.timestamp() as f64;
+
+        // We round to 2 digits in order to avoid high-precision sample rates that are more difficult
+        // to work with. Rounding is performed following the nearest integer.
+        let sample_rate_difference =
+            f64::round((from_sample_rate - self.external_params.base_sample_rate) * 100.0) / 100.0;
```

Comment:
It is easier to debug and in general I thought we were going to keep 2 decimal digits.

Reply:
By multiplying `sample_rate_difference * inverse_progress`, you potentially get a lot more decimals again, right? If we want to do 1% steps of decay (e.g. `0.7, 0.69, 0.68, ...`) I would round when the computation is done, e.g. in `get_sample_rate`.

Reply:
What you are saying is correct but I did that on purpose.

I had at first rounded only the sample difference because it was behaving quite weird, in the sense that I was getting 0.7 - 0.2 = 0.4999999 due to floating arithmetic approximations I think. And this was going to screw up the whole calculation because we started doing that with an ill number.

I think that we could simplify everything and round at the end, with some more loss propagated through multiplications but that is fine. The only problem I see is that we will have weird numbers in tests if we choose for example the 50% progress time.

Reply:
Do increased precision and floating point glitches really make this more difficult to work with? In the end, you can just make your "correct" calculation and use the result. If you ever have to debug print the value and receive `0.49999...`, it is still clear what the value is. Besides, interpolation will give you all sorts of values anyway. 

Also, please do not round to integer percentages. We need more precision than that, considering that a realistic uniform sample rate is 1%. You can, of course, round your rule targets, but please do not round here.

I would suggest to remove complexity and not round anywhere in Relay.

Reply:
Great! Thanks for the explanation Jan!